Feature: Add Command Menu to mobile devices

fbmcipher opened the pull request:
As part of our work towards optimising the UI for mobile and first-time users, this PR adds a Command Menu to **em**.

Note: The Command Menu is only available on mobile devices. It can be tested in desktop Chrome using the Responsive Design mode.

## Key points
* Adds a "..." menu to the cursor thought.
    * As discussed, this isn't animated right now.
* When the "..." is tapped, a new *panel* slides up from the bottom of the screen displaying a set of common commands that the user can execute on the cursor thought thought.
* The user can execute thoughts from the panel.
* They can still continue to navigate their thoughtspace on the top half of the screen – including scrolling and changing the cursor thought.
* The user can close the Command Panel by swiping it down with their finger, or tapping the "X" icon in the corner.

## Other notes
* This iteration of the Command Menu contains a **static set of commands** – though the layout and selection of these commands is very easily modified by editing [`PanelCommandGrid.tsx`](https://github.com/tangerine-studio-ltd/em/blob/command-menu/src/components/CommandMenu/PanelCommandGrid.tsx).
* This iteration of the Command Menu contains only the **first screen** – in other words, you cannot **expand the panel further to see all commands**.
    * We will revisit these features at a later time.
* Care has been applied to ensure the panel interacts gracefully with the on-screen keyboard. If the user does something that would open the on-screen keyboard, the panel closes itself to make way for the keyboard – and vice versa. This allows for a very natural flow between editing and executing commands.
* The elipsis menu doesn't show during the tutorial, as that would probably break the tutorial flow. At some point we should probably revisit the tutorial and think about how to integrate the new UI elements into it – though that is a problem for a later time.

Looking forward to your thoughts, @raineorshine!

---

**Thanks:** @pierogizwigilii

raineorshine commented on src/components/Note.tsx:
```diff
@@ -97,6 +98,13 @@ const Note = React.memo(({ path }: { path: Path }) => {
   const onFocus = () => {
+    const state = store.getState()
+
+    // Close command menu if it's open when focusing on a note.
+    if (state.showCommandMenu && isTouch) {
+      dispatch(toggleCommandMenuActionCreator({ value: false }))
+    }
```
Let's close the Command Menu in the `setCursor` reducer whenever `editing` is set to true. That will cover this case and others without having to dispatch another action, and the logic will be a bit more hidden away.

fbmcipher replied:
Nice, that does feel cleaner. Resolved in 9a8cd20ce34bff43b533c839800f7bf637529846.

raineorshine commented on src/components/Editable.tsx:
```diff
@@ -496,9 +497,14 @@ const Editable = ({
       if (!dragHold && !dragInProgress) {
         setCursorOnThought({ editing: true })
+
+        // Close command menu when entering edit mode on a touch device
+        if (showCommandMenu && isTouch) {
+          dispatch(toggleCommandMenuActionCreator({ value: false }))
+        }
```
This will be covered if we move this logic to the `setCursor` reducer.

See other comment.

raineorshine commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,53 @@
+        }),
+      )}
+    >
+      {React.Children.map(children, (child, index) => {
+        if (React.isValidElement(child)) {
+          return React.cloneElement(child, {
```
Interesting approach! I learned something new.

raineorshine commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,53 @@
+import { css, cx } from '../../../styled-system/css'
+import { panelCommandGroupRecipe } from '../../../styled-system/recipes'
+
+interface PanelCommandGroupProps {
+  children: React.ReactNode
+}
```
React comes with a helper type for this:

```diff
diff --git a/src/components/CommandMenu/PanelCommandGroup.tsx b/src/components/CommandMenu/PanelCommandGroup.tsx
index 5ccdb089a2a..680fb8f4c33 100644
--- a/src/components/CommandMenu/PanelCommandGroup.tsx
+++ b/src/components/CommandMenu/PanelCommandGroup.tsx
@@ -1,13 +1,9 @@
-import React from 'react'
+import React, { PropsWithChildren } from 'react'
 import { css, cx } from '../../../styled-system/css'
 import { panelCommandGroupRecipe } from '../../../styled-system/recipes'
 
-interface PanelCommandGroupProps {
-  children: React.ReactNode
-}
-
 /** A component that groups two PanelCommand components together. */
-const PanelCommandGroup: React.FC<PanelCommandGroupProps> = ({ children }) => {
+const PanelCommandGroup: React.FC<PropsWithChildren> = ({ children }) => {
   const childCount = React.Children.count(children)
 
   // Determine the size from the first child
```

fbmcipher replied:
Good to know! Thanks, I've swapped that out now in 57603048e8274de2edfe7db4e25b578c7e816802.

raineorshine commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,53 @@
+                  gridColumn: 'span 1 !important',
+                }),
+            }),
+          } as any)
```
We can typecast the `child`, which is a bit more type safe than using `any`:

```diff
diff --git a/src/components/CommandMenu/PanelCommandGroup.tsx b/src/components/CommandMenu/PanelCommandGroup.tsx
index c1e64bd5823..5ccdb089a2a 100644
--- a/src/components/CommandMenu/PanelCommandGroup.tsx
+++ b/src/components/CommandMenu/PanelCommandGroup.tsx
@@ -31,7 +31,7 @@ const PanelCommandGroup: React.FC<PanelCommandGroupProps> = ({ children }) => {
     >
       {React.Children.map(children, (child, index) => {
         if (React.isValidElement(child)) {
-          return React.cloneElement(child, {
+          return React.cloneElement(child as React.ReactElement, {
             className: css({
               borderRadius: index === 0 ? '16px 0 0 16px' : index === childCount - 1 ? '0 16px 16px 0' : '0',
               flex: 1,
@@ -42,7 +42,7 @@ const PanelCommandGroup: React.FC<PanelCommandGroupProps> = ({ children }) => {
                   gridColumn: 'span 1 !important',
                 }),
             }),
-          } as any)
+          })
         }
         return null
       })}
```

fbmcipher replied:
Good eye! Cleaned up in f5835c4fcd6e1b5037d4e35f2d424be3c690d991.

raineorshine reviewed (changes requested):
Thank you for the submission! Looks great.

---

Let's also close the Command Menu when clicking empty space in the editor, which already closes the virtual keyboard.

```diff
diff --git a/src/components/Content.tsx b/src/components/Content.tsx
index ebfc94ef497..952478c8c1b 100644
--- a/src/components/Content.tsx
+++ b/src/components/Content.tsx
@@ -7,6 +7,7 @@ import { Thunk } from '../@types/Thunk'
 import { closeModalActionCreator as closeModal } from '../actions/closeModal'
 import { expandContextThoughtActionCreator as expandContextThought } from '../actions/expandContextThought'
 import { toggleColorPickerActionCreator as toggleColorPicker } from '../actions/toggleColorPicker'
+import { toggleCommandMenuActionCreator as toggleCommandMenu } from '../actions/toggleCommandMenu'
 import { toggleLetterCaseActionCreator as toggleLetterCase } from '../actions/toggleLetterCase'
 import { isTouch } from '../browser'
 import { ABSOLUTE_PATH, HOME_PATH, TUTORIAL2_STEP_SUCCESS } from '../constants'
@@ -61,6 +62,7 @@ const Content: FC = () => {
       state.showModal ? closeModal() : null,
       state.expandedContextThought && !state.noteFocus ? expandContextThought(null) : null,
       state.showColorPicker ? toggleColorPicker({ value: false }) : null,
+      state.showCommandMenu ? toggleCommandMenu({ value: false }) : null,
       state.showLetterCase ? toggleLetterCase({ value: false }) : null,
     ])
   }

```

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,98 @@
+interface PanelCommandProps {
+  command: Command
+  size?: 'small' | 'medium' | 'large' | 'xlarge'
+  className?: string
```
Good catch – this was leftover from a previous implementation of PanelCommand + PanelCommandGroup and has now been removed.

fbmcipher replied:
We use `className` in `PanelCommandGroup` on the cloned `PanelCommand` to modify the appearance of the commands, giving them that conjoined appearance.

https://github.com/tangerine-studio-ltd/em/blob/e857d2eba6428e6534bd8d5ad643ac164b2ff1a3/src/components/CommandMenu/PanelCommandGroup.tsx#L35-L44

raineorshine replied:
Ah, missed that! Thank you.

Generally I would like to avoid props like `className` or `cssRaw` which provide a very wide API. It creates a tight coupling between the parent and child, and lacks type safety. For example, the border-radius of 16px is spread across both `PanelCommandGroup` and `PanelCommand`. Why does `PanelCommandGroup need to know the border-radius of its children? Actually, it just needs to control its position within the group. A prop like `{ position?: 'first' | 'last' }` would allow the border-radius to be defined in one place, and narrow the API between the parent and child.

fbmcipher replied:
Good observations and rationale!

We address this in 50d67752f1054f35a966b5e87f329c0f6be8f803 with a new `groupPosition` prop, and use a PandaCSS recipe to declare variants for `PanelCommand`.

While I'm at it, I move all of `PanelCommand`'s styles into the recipe for the sake of consistency – and create variants for the other conditional styles that `PanelCommand` had (i.e. `isButtonExecutable`, `isButtonActive`, and `size`)

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,98 @@
+        }),
+        className ? className : css({}),
+      )}
+      title={`${command.label}${command.keyboard ? ` (${formatKeyboardShortcut(command.keyboard)})` : ''}`}
```
Got it – thanks! Removed.

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,98 @@
+        animated={isAnimated}
+        animationComplete={() => setIsAnimated(false)}
+      />
+      {command.id !== 'indent' && command.id !== 'outdent' && (
```
Rather than hardcoding, how about we add a property in the `Command` interface that controls whether or not we hide/show the label?

See: c8b70f48d38be09713f52d20534e6c7c218f075e

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,98 @@
+          borderRadius: '16px',
+          backgroundColor: isButtonActive ? 'purple' : 'gray15',
+          cursor: isButtonExecutable ? 'pointer' : 'default',
+          transition: 'opacity 0.5s ease, background-color 0.5s ease',
```
Thanks for flagging – this has been resolved in 5e4de9b3e0dc6b34ca163614046043dfcf844974.

fbmcipher commented on src/components/CommandMenu/PanelCommandGrid.tsx:
```diff
@@ -0,0 +1,47 @@
+        maxWidth: '100%',
+      })}
+    >
+      <PanelCommandGroup>
+        <PanelCommand command={{ ...toggleDone, label: 'Done' }} size='small' />
+        <PanelCommand command={note} size='small' />
+        <PanelCommand command={{ ...favorite, label: 'Favorite' }} size='small' />
+        <PanelCommand command={deleteCommand} size='small' />
+      </PanelCommandGroup>
```
Thank you! 🙏 Clean composition was a design goal for this particular project, so I'm glad that's come through in the end result.

fbmcipher commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,166 @@
+        // Disable swipe to open - this removes the swipe-up-to-open functionality
+        disableSwipeToOpen={true}
+        ref={containerRef}
+        transitionDuration={500}
```
Thanks for picking up on this – fixed in 5e4de9b3e0dc6b34ca163614046043dfcf844974.

fbmcipher commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,166 @@
+  /** Toggle the command menu. */
+  const toggleCommandMenu = (value: boolean) => {
+    dispatch(toggleCommandMenuActionCreator({ value }))
+  }
```
Ah, you're right. Thanks for catching.

I've aliased `toggleCommandMenu` as advised and removed the extraneous function in f6ffcff83081f9c7dbc34d1856987b4f3ccce33d. Thanks!

fbmcipher commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,166 @@
+          )}
+          onTouchEnd={() => {
+            setIsSwiping(false)
+          }}
```
Good catch – agreed, looks like it's unused.. Removed in 087cc59652cef27358eeb67d454fab1b1dbef3a3.

fbmcipher commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,166 @@
+        // Use PaperProps to directly target the Paper component
+        PaperProps={{
+          style: {
+            backgroundColor: token('colors.darkgray'),
+            // Make sure it overrides any inline styles
+            display: 'flex',
+            flexDirection: 'column',
+            width: '100%',
+            overflow: 'hidden',
+            maxHeight: '70%',
+            pointerEvents: 'auto',
```
Ah, it looks like the styles in `classes` above this were from a previous attempt to style the menu.

MUI can be a little difficult about how you style it. I tried pulling out the styled in `PaperProps` to see if the styles in `classes` would still apply – but they didn't.

So I removed the extraneous styles from `classes` and just stick to using `PaperProps` and `ModalProps`, which IIRC is the MUI-sanctioned way of styling their components.

ed75ac32c69ef0241d6feee3cb94808380e6a590

fbmcipher commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,166 @@
+  const prevCursorRef = useRef(cursor)
+
+  useEffect(() => {
+    // Only close the command menu if cursor becomes inactive but make sure it doesn't close if the cursor is just switching
```
Sure.

This code technically handles _two_ things. The first of which is: **when the cursor becomes inactive (i.e. no thought is focussed), close the command menu.** It does that by setting up a `useEffect` which observes the `cursor` changing, evaluates if the `cursor` is null, and if it is dispatches the close action.

However – there was a bug with this – when the cursor changes _between_ thoughts, the panel got closed. So in some time between the cursor going from A to B, the `cursor` evaluates as falsey – and the panel gets closed.

To mitigate that, there's a 200ms timeout – allowing the cursor state to change from its old state to its new state – at which point we evaluate. This _does_ fix the bug, but perhaps there's a way to tidy up this logic? Do you have any idea what might be happening based on my description?

raineorshine replied:
Okay, thanks for the explanation.

I think that at a high level, a side effect is the wrong place so do this. Here we have a change in Redux state (`cursor`) that results in a change in Redux state (`showCommandMenu`). If we were reacting to browser event or effecting a browser change, then a component hook would make sense. However this seems like a pure Redux state change, and should be handled in a reducer if possible, such as [setCursor](https://github.com/cybersemics/em/blob/6b63424684c233e4a5169277f47130616f8deec8/src/actions/setCursor.ts). If it cuts across several reducers, we can update the state in a store enhancer or Redux middleware.

> However – there was a bug with this – when the cursor changes _between_ thoughts, the panel got closed. So in some time between the cursor going from A to B, the `cursor` evaluates as falsey – and the panel gets closed.

The `cursor` is not falsey when changing from one thought to another. I think there's something else going on here. I'm skeptical that we need a side effect and a ref to achieve this functionality. See if closing the command menu in the `setCursor` reducer works. If not, there still may be a reducer-only approach.

fbmcipher replied:
> The cursor is not falsey when changing from one thought to another. I think there's something else going on here. I'm skeptical that we need a side effect and a ref to achieve this functionality. See if closing the command menu in the setCursor reducer works. If not, there still may be a reducer-only approach.

Thank you for this pointer – yes, that makes sense, and it does work this way

In fa27159ed3b7c1fda7295460793a96a7be607455 and 31eb4142768e167866e0abb68effa3c2dd83bc7f, we now close the command menu directly in the setCursor reducer rather than dispatching actions with effects. I see that this keeps all of the state inside Redux, which is much cleaner – keeping all of the logic in one place makes it more predictable and less brittle.

This fix actually exposed a bug where `onClick` events to the elipsis icon bubbled up to Content.clickOnEmptySpace – so soon as I tap the elipsis, the menu opens and quickly closes again due to Content.clickOnEmptySpace. That is fixed in 589c7927a0819eb0f08248fd994bf82d2f1d02cf – and we now use the `fastClick` util instead of `onClick` for consistency and better responsiveness on mobile.

I believe that addresses this point! Thank you for your notes.

fbmcipher commented:
Thanks for reviewing @raineorshine!

* I've applied your diff to close the Command Menu when you tap empty space.
* I've removed `subcategoriseAll` from the Command Grid for now. That does leave a 2x1 space open in the grid – do you have any ideas on what command(s) you'd like to place there instead?
* I address a z-index oversight where the Command Menu displayed above the Sidebar.
* I've addressed all the comments raised above – please respond and re-review when you can!

Thank you very much 🙌

raineorshine commented on src/actions/setCursor.ts:
```diff
@@ -163,7 +164,12 @@ const setCursor = (
+  dispatch => {
+    // If editing is being set to true, close the command menu first
+    if (payload.editing === true) {
+      dispatch(toggleCommandMenuActionCreator({ value: false }))
```
It would be preferable to push this all the way down to the reducer. Action-creators are better for browser state and side effects, while reducers are better for pure input/output.

Followup on https://github.com/cybersemics/em/pull/2874#discussion_r2019971281.

raineorshine reviewed (changes requested):
Thanks for addressing all the comments! It was helpful to see the commits.

> * I've removed `subcategoriseAll` from the Command Grid for now. That does leave a 2x1 space open in the grid – do you have any ideas on what command(s) you'd like to place there instead?

The counterpart to `categorize` (formerly `subcategorizeOne`) is `collapseContext`, so that would be a natural choice. 

Also, I've been wanting to rename `collapseContext` since "collapse" brings to mind expanding/collapsing bullets. What do you think of `uncategorize` to make the inverse relationship with `categorize` super clear?

fbmcipher reviewed (commented):
Hey @raineorshine – I've addressed the remainder of your points in the last few commits. Thank you very much for the detailed feedback as always.

Additionally, I've merged with `main`, updated our code in light of the `LayoutTree` refactor, and replaced `subcategorizeOne` with the `categorize` command.

trevinhofmann commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,48 @@
+        React.Children.map(children, (child, index) => {
+          if (React.isValidElement(child)) {
+            return React.cloneElement(child as React.ReactElement, {
+              groupPosition: index === 0 ? 'first' : index === childCount - 1 ? 'last' : 'between',
```
Instead of implementing this logic, would it be possible to make use CSS pseudo-class selectors such as [`:first-of-type`](https://developer.mozilla.org/en-US/docs/Web/CSS/:first-of-type) and [`:last-of-type`](https://developer.mozilla.org/en-US/docs/Web/CSS/:last-of-type)?

fbmcipher replied:
Wow – you're right! Looking back, the implementation here is certainly overcomplicated.

Refactored in bd74d4f93939683544475fe5c29d75ca6d816df1 – we just use the CSS pseudo-class selectors, which works just as well and allows us to strip out the fragile `React.Children` implementation.

trevinhofmann reviewed (changes requested):
Thank you for the pull request, @fbmcipher! This is a great feature!

Is there an issue number that this resolves? I wasn't able to find one.

---

Is the layout meant to be an aligned grid? In this screenshot, the gap between `Done` and `Note` is slightly misaligned with the gap between `Outdent` and `Indent`.

<img width="428" alt="Screenshot 2025-04-13 at 11 49 19 AM" src="https://github.com/user-attachments/assets/1d3f6698-a260-40f7-8605-e119d8396767" />

---

## 1. The `Categorize` command does not start editing

### Steps to reproduce

```
- Foo
```

1. Open the command menu on `Foo`
2. Click the `Categorize` command

### Current Behavior

The command menu is closed, and the new thought is not being edited / the keyboard does not open.

### Expected Behavior

The keyboard should open and allow the user to begin editing the newly-created thought.

---

## 1. The `Note` command does not start editing

### Steps to reproduce

```
- Foo
```

1. Open the command menu on `Foo`
2. Click the `Note` command

### Current Behavior

The command menu is closed, and the new note is not being edited / the keyboard does not open.

### Expected Behavior

The keyboard should open and allow the user to begin editing the newly-created note.

raineorshine commented:
Thank you @trevinhofmann!

@fbmcipher I recommend looking at how Toolbar handles keyboard/selection behavior on tap, as I believe the problem has been solved there.

fbmcipher commented on src/actions/toggleCommandMenu.ts:
```diff
@@ -0,0 +1,17 @@
+/** Toggles the Command Menu. */
+const toggleCommandMenu = (state: State, { value }: { value?: boolean } = {}) => ({
+  ...state,
+  showCommandMenu: value == null ? !state.showCommandMenu : value,
```
Nice catch! Hopefully 995b11e30e6df6074b0d0ab3182195b9cf0945d4 makes this clearer.

fbmcipher commented on src/actions/setCursor.ts:
```diff
@@ -155,6 +155,8 @@ const setCursor = (
       : null),
     ...(!thoughtsResolved ? { showColorPicker: false, showLetterCase: false, showSortPicker: false } : null),
+    // Close command menu when editing is set to true, or if there is no cursor.
+    showCommandMenu: editing === true || !thoughtsResolved ? false : state.showCommandMenu,
```
Yeah, the way you've written this is much clearer! Thanks. 755d86858b735b9f4dc2da46f9fb063ce640f41c takes care of this.

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,79 @@
+  if (!command) {
+    console.error('Missing command')
+    return null
+  }
```
You're right, we don't need this explicit check. I've removed it in 04a66bf63b7e77dabca013be94f17e0709887dc6.

fbmcipher commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,79 @@
+  const commandState = useSelector(state => isActive?.(state))
+  const isButtonActive = commandState
+
+  const SVG = svg as React.FC<Icon>
```
Ah, thanks for the heads up.

We handle this better in 6cbc98c5c8457369e712ed3a63e653e01bddefce:

* We don't use the `as` type assertion unless `svg` is defined.
* We don't render an icon in the PanelCommand if we don't have one.

fbmcipher commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,48 @@
+
+/** A component that groups two PanelCommand components together. */
+const PanelCommandGroup: React.FC<PropsWithChildren> = ({ children }) => {
+  const childCount = React.Children.count(children)
```
Addressed below – we were actually able to pull out the `React.Children`-based implementation and use pseudo selectors instead! Thanks again for the great. catch.

fbmcipher commented:
Thank you very much for the review, @trevinhofmann!

I've addressed the code quality concerns and also fixed the issues you found.

> Is the layout meant to be an aligned grid? In this screenshot, the gap between Done and Note is slightly misaligned with the gap between Outdent and Indent.

The Done/Note/Favourite/Delete commands are actually *not* meant to be in a group! Groups are only meant to be used for connected commands (like Indent/Outdent) – so is designed for up to two commands.

I pulled them out of the group in dd3886307c36cd40132a46d799f83025cc943d40, and now the commands align as expected.

![em](https://github.com/user-attachments/assets/737637e0-4e67-48b6-836c-925fb05bc374)

> 1. The Categorize command does not start editing

> 2. The Note command does not start editing

Fixed in d9cc5499ba814287cbb0f983e4d26b4556dd3b31 - this was happening on Android, but not on iOS. Like in `ToolbarButton`, we needed to preventDefault on the event to prevent the Editable component from being blurred. Thanks @raineorshine for the pointer to look in Toolbar!

> Is there an issue number that this resolves? I wasn't able to find one.

There isn't an issue associated with this PR, as the work was arranged directly between @raineorshine and I.

Future mobile UI work like this will be managed through GitHub Projects, so future PRs will have issues attached to them, just like the rest of the project.

trevinhofmann commented on src/components/BulletEllipsis.tsx:
```diff
@@ -0,0 +1,32 @@
+  const dispatch = useDispatch()
+  return (
+    <div
+      {...fastClick(() => dispatch(toggleCommandMenuActionCreator()))}
```
We have another PR open which will replace `fastClick`: https://github.com/cybersemics/em/pull/2886

I would recommend proactively replacing these new usages of `fastClick` in this PR to avoid conflicts. Generally, the replacement is to:
1. Use `onClick` instead
2. Use `role='button'`
3. Call `haptics.light()` if applicable

fbmcipher replied:
Thanks for letting me know!

Given Raine's message in https://github.com/cybersemics/em/pull/2886#pullrequestreview-2793497052, I'll leave these fastClick usages for now as it seems it's sticking around.

trevinhofmann commented on src/components/CommandMenu/PanelCommand.tsx:
```diff
@@ -0,0 +1,81 @@
+        isButtonExecutable,
+        isButtonActive,
+      })}
+      {...fastClick(handleTap)}
```
Another case of `fastClick` to replace.

fbmcipher replied:
Thanks for letting me know!

Given Raine's message in https://github.com/cybersemics/em/pull/2886#pullrequestreview-2793497052, I'll leave these fastClick usages for now as it seems it's sticking around.

trevinhofmann commented on src/components/CommandMenu/PanelCommandGroup.tsx:
```diff
@@ -0,0 +1,34 @@
+
+/** A component that groups two PanelCommand components together. */
+const PanelCommandGroup: React.FC<PropsWithChildren> = ({ children }) => {
+  const childCount = React.Children.count(children)
```
Still using `React.Children` with the latest implementation. I won't push back further on removing it, but perhaps a simpler implementation could be to pass the size to the `PanelCommandGroup` directly rather than trying to infer it?

fbmcipher replied:
I do understand the concern with `React.Children`. In 9af6faa73b2706cd99d82aaab4aa75a4e8051a80, we now pass the size and child count directly, allowing us to fully pull out `React.Children`.

We lose some of the "magic" of being able to simply drop commands into a group, but that's not the biggest loss.

trevinhofmann reviewed (changes requested):
Thank you for the updates, @fbmcipher! Apologies for the delay in reviewing; I've been out sick for a few days.

(1) & (2) have been resolved 🎉

I'm noting (3) below, but it may be outside the scope of this PR since it seems to be an issue with the existing `Uncategorize` command rather than anything introduced here. Please let me know your thoughts on this, otherwise the `fastClick` replacement is the only issue I see remaining.

---

## 3. The `Uncategorize` button remains enabled

### Steps to reproduce:

```
- Foo
```

1. Open the command menu on `Foo`
2. Click `Uncategorize`

### Current behavior:

The `Uncategorize` button is enabled and animates without doing anything.

### Expected behavior

Like the `Swap Parent` button, the `Uncategorize` button should be disabled when no action would be done (when there are no children).

fbmcipher commented:
Thanks @trevinhofmann – hope you're feeling better!

I believe that's everything resolved:

* I took a quick look at the issue with `uncategorize`, and it does seem to be an issue with the command itself, rather than an issue with this command menu, so I think it's out of scope fort his PR. There may be scope to open a new issue here.
* Thanks for raising the flag on `fastClick`, I appreciate the proactivity. It looks like a decision has been made to keep the `fastClick` helper in https://github.com/cybersemics/em/pull/2886#pullrequestreview-2793497052 – so I've left them in for now. Just let me know if they do indeed need to change, and I'd be happy to make the necessary changes for this PR.
* I pulled out the last remaining use of `React.Children`.

trevinhofmann commented:
Puppeteer tests were failing. I'm rerunning the check now in case it's due to flaky tests.

trevinhofmann commented:
@fbmcipher - Looks like just a flaky test issue 👍

There are merge conflicts, though, and I don't have permission to push to the branch to resolve them. Can you please have a look?

fbmcipher commented:
Thanks so much @trevinhofmann!

I've resolved the merge conflicts, and made a couple minor changes d2e29ca and 3c54865 to adapt to API changes made in PRs #2891 and #2911.

Should be ready to merge now! 🙌

raineorshine reviewed (changes requested):
Thanks!

raineorshine commented:
There is an expected merge conflict from #2916 that needs to be resolved now. Sorry for the trouble! I don't have push access to this PR, otherwise I would have resolved it myself since this is so close to being merged.

In #2916, we've consolidated the logic to toggle menus/dropdowns since usually they are mutually exclusive. The merge conflict resolution should go something like this:

- Add `showCommandMenu` to the appropriate places in `toggleDropdown`.
- You may be able to move `toggleCommandMenu` completely into `toggleDropdown`.
- In `Content.tsx`, use the upstream change.

fbmcipher commented on src/components/BulletEllipsis.tsx:
```diff
@@ -0,0 +1,31 @@
+        left: -56,
+        top: 3,
+        width: 30,
+        height: 30,
```
In 38f8d5af4c946333ea5235008e569e478d1b0379, we now scale the BulletEllipsis with the in-app font size setting. 

Just using em units wasn't enough to the ellipsis positioned nicely and relatively alongside the Bullet, so I moved the ellipsis to inside the Bullet and set some styles to make sure the ellipsis is *always* a consistent distance away from the Bullet, no matter the font size.

fbmcipher commented on src/actions/toggleCommandMenu.ts:
```diff
@@ -0,0 +1,31 @@
+  if (document && document.activeElement) {
+    const activeEl = document.activeElement as HTMLElement
+    activeEl?.blur?.()
+  }
```
It looks like it's unnecessary. I've removed it in a3f065f24b732d95a3904372798e351344680164, which makes the action much cleaner.

raineorshine commented on src/components/CommandMenu/CommandMenuPanel.tsx:
```diff
@@ -0,0 +1,87 @@
+        data-testid='command-menu-panel'
+        // Disable swipe to open - this removes the swipe-up-to-open functionality
+        disableSwipeToOpen={true}
+        ref={containerRef}
```
Looks like `containerRef` is no longer used and can be removed. Leftover from `Sidebar` code.
